os/bluestore: Refactor Bluestore Caches

markhpc opened the pull request:
This PR currently includes both the trim on add commit from wip-bs-cache-evict (#28578) and also a new commit that splits the onode and buffer caches into their own distinct entities with independent locking.  This avoids having onode add/lookup and buffer eviction contend on the same cache shard lock and provides a moderate performance improvement beyond the trim-on-add improvement in wip-bs-cache-evict.  This PR works in conjunction with #27705 and provides a significant performance improvement over master with potential room for further improvement.

![FIO+LibRBD 4K RandWrite (Single-OSD) LRU Cache](https://user-images.githubusercontent.com/1286295/59611145-68fd1000-90e0-11e9-8449-ce40145a619f.png)

Average and tail Latency also decreases, though tail latency increases slightly with the addition of #27705 (though average latency continues to decrease and IOPS increases).


msec | Master Latency | wip-bs-cache-evict | wip-bs-ind-cache | wip-bs-ind-cache + PR #27705
-- | -- | -- | -- | --
min | 0.305 | 0.308 | 0.287 | 0.297
avg | 8.934625 | 7.094215 | 6.49727 | 5.92275
99.9th percentile | 86 | 71 | 93 | 116
99.99th percentile | 334 | 113 | 112 | 232
max | 1142.1 | 571.779 | 556.056 | 702.879

Surprisingly, increasing the number of onode and buffer cache shards from 1 to 4 increased lock contention and hurt performance by several percent.  More analysis will be necessary to understand why this is the case.  Other areas for improvement may include reducing lock contention due to cache stat updates in the mempool thread.

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

liewegas commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -988,6 +973,21 @@ void BlueStore::LRUCache::_trim(uint64_t onode_max, uint64_t buffer_max)
   }
 }
 
+void BlueStore::LRUCache::_trim_buffers_to(uint64_t max) {
```
nit: open brace goes on next line in .cc files

markhpc replied:
fixed!

markhpc commented on src/os/bluestore/BlueStore.h:
```diff
@@ -1084,38 +1086,137 @@ class BlueStore : public ObjectStore,
+
+  /// A Generic buffer Cache Shard
+  struct BufferCacheShard : public CacheShard {
+    std::array<std::pair<ghobject_t, mono_clock::time_point>, 64> dumped_onodes;
```
Seems a little shady to me to leave this in the BufferCacheShard.  Should go back and look at that again.

markhpc replied:
Also, I wonder if "Buffer Cache" is still the right name for this.

liewegas replied:
i think 'buffer' is still accurate

markhpc replied:
Moved!

tchaikov commented:
http://pulpito.ceph.com/kchai-2019-06-20_05:33:07-rados-wip-kefu-testing-2019-06-20-1103-distro-basic-smithi/

the valgrind failures are caused by #27834

tchaikov commented:
http://pulpito.ceph.com/kchai-2019-06-20_05:33:07-rados-wip-kefu-testing-2019-06-20-1103-distro-basic-smithi/

the valgrind failures are caused by #27834

ifed01 commented on src/os/bluestore/BlueStore.h:
```diff
@@ -1086,276 +1086,281 @@ class BlueStore : public ObjectStore,
-      onode_max = max;
+    void add_stats(uint64_t *onodes) override {
+      std::lock_guard l(lock);
+      *onodes += lru.size();
```
may be start tracking  lru size using atomic_int and get rid off lock here?

markhpc replied:
Yep, I agree.  Done!

ifed01 commented on src/os/bluestore/BlueStore.h:
```diff
@@ -1086,276 +1086,281 @@ class BlueStore : public ObjectStore,
+                   uint64_t *blobs,
+                   uint64_t *buffers,
+                   uint64_t *bytes) override {
       std::lock_guard l(lock);
```
similarly we can get rid off the lock if tracking lru size externally using atomic

markhpc replied:
Agreed. Done!

markhpc commented on src/os/bluestore/BlueStore.h:
```diff
@@ -1086,276 +1086,281 @@ class BlueStore : public ObjectStore,
-      buffer_max = max;
+  /// A Generic buffer Cache Shard
+  struct BufferCacheShard : public CacheShard {
+    std::array<std::pair<ghobject_t, mono_clock::time_point>, 64> dumped_onodes;
```
Done!

markhpc commented:
New changes:

nits cleanup
moved dumped_onodes into the onode cache
external tracking of cache size using an atomic to avoid lock for stats
removed trim from discard and only trim when we lock (address Igor's previous comments)

Performance didn't appear to change significantly:

wip-bs-ind-cache:

before: 39564 IOPS
after: 39008 IOPS

wip-bs-ind-cache + #27705:

before: 43219 IOPS
after: 43286 IOPS

Inverse wallclock profiles attached for the base and with #22705 cases.  There's still some lock contention on the cache in the tp_osd_tp threads (#22705 helps by avoiding double caching) but now we're hitting pglock as well.  Bigger fish to fry may be memcpy in BlueRocksWritableFile::Append.
[wip-bs-ind-cache-update.inverse.txt](https://github.com/ceph/ceph/files/3322840/wip-bs-ind-cache-update.inverse.txt)
[wip-bs-ind-cache-update-27705.inverse.txt](https://github.com/ceph/ceph/files/3322841/wip-bs-ind-cache-update-27705.inverse.txt)

markhpc commented:
Looking at:

http://pulpito.ceph.com/sage-2019-06-26_20:55:20-rados-wip-sage-testing-2019-06-26-0958-distro-basic-smithi/

I see a ton of failures due to ceph-objecstore-tool being missing.  Possibly several with scrubbing terminating with not all pgs active/clean.

tchaikov commented on src/os/bluestore/BlueStore.h:
```diff
@@ -1131,208 +1235,137 @@ class BlueStore : public ObjectStore,
-	boost::intrusive::list_member_hook<>,
-	&Onode::lru_item> > onode_lru_list_t;
+  /// A TwoQ buffer Cache Shard
+  struct TwoQBufferCacheShard : public BufferCacheShard {
```
can we move the different implementations of `BufferCacheShard` into .cc ? as they are not part of the interface of BlueStore.

tchaikov commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -13707,8 +13633,12 @@ void BlueStore::generate_db_histogram(Formatter *f)
   dout(10) << __func__ << dendl;
-  for (auto i : cache_shards) {
-    i->trim_all();
+  for (auto i : onode_cache_shards) {
```
```
2019-07-09T09:24:36.061 INFO:teuthology.orchestra.run.mira052.stderr:/build/ceph-15.0.0-2580-g98c02be/src/os/bluestore/BlueStore.cc: In function 'virtual void BlueStore::LruOnodeCacheShard::_trim_to(uint64_t)' thread 7f7db9edbb40 time 2019-07-09T09:24:35.970556+0000
2019-07-09T09:24:36.061 INFO:teuthology.orchestra.run.mira052.stderr:/build/ceph-15.0.0-2580-g98c02be/src/os/bluestore/BlueStore.cc: 925: FAILED ceph_assert(num == 1)
2019-07-09T09:24:36.061 INFO:teuthology.orchestra.run.mira052.stderr: ceph version 15.0.0-2580-g98c02be (98c02be3ad8838c698222ff1581d0d7e4d9b58a1) octopus (dev)
2019-07-09T09:24:36.061 INFO:teuthology.orchestra.run.mira052.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x152) [0x7f7dafd0e80a]
2019-07-09T09:24:36.061 INFO:teuthology.orchestra.run.mira052.stderr: 2: (ceph::__ceph_assertf_fail(char const*, char const*, int, char const*, char const*, ...)+0) [0x7f7dafd0e9e5]
2019-07-09T09:24:36.061 INFO:teuthology.orchestra.run.mira052.stderr: 3: ceph-objectstore-tool() [0x1001c3d]
2019-07-09T09:24:36.062 INFO:teuthology.orchestra.run.mira052.stderr: 4: (BlueStore::_flush_cache()+0x80) [0x1061940]
2019-07-09T09:24:36.062 INFO:teuthology.orchestra.run.mira052.stderr: 5: (BlueStore::_fsck(bool, bool)+0x4ad9) [0x107b559]
2019-07-09T09:24:36.062 INFO:teuthology.orchestra.run.mira052.stderr: 6: (BlueStore::_mount(bool, bool)+0x20a) [0x1083aaa]
2019-07-09T09:24:36.062 INFO:teuthology.orchestra.run.mira052.stderr: 7: (main()+0x2dfe) [0xadc5fe]
2019-07-09T09:24:36.062 INFO:teuthology.orchestra.run.mira052.stderr: 8: (__libc_start_main()+0xf0) [0x7f7daeb15830]
2019-07-09T09:24:36.062 INFO:teuthology.orchestra.run.mira052.stderr: 9: (_start()+0x29) [0xaf21e9]
2019-07-09T09:24:36.062 INFO:teuthology.orchestra.run.mira052.stderr:*** Caught signal (Aborted) **
2019-07-09T09:24:36.063 INFO:teuthology.orchestra.run.mira052.stderr: in thread 7f7db9edbb40 thread_name:ceph-objectstor
2019-07-09T09:24:36.063 INFO:teuthology.orchestra.run.mira052.stderr: ceph version 15.0.0-2580-g98c02be (98c02be3ad8838c698222ff1581d0d7e4d9b58a1) octopus (dev)
2019-07-09T09:24:36.063 INFO:teuthology.orchestra.run.mira052.stderr: 1: (()+0x11390) [0x7f7daf3e1390]
2019-07-09T09:24:36.063 INFO:teuthology.orchestra.run.mira052.stderr: 2: (gsignal()+0x38) [0x7f7daeb2a428]
2019-07-09T09:24:36.063 INFO:teuthology.orchestra.run.mira052.stderr: 3: (abort()+0x16a) [0x7f7daeb2c02a]
2019-07-09T09:24:36.063 INFO:teuthology.orchestra.run.mira052.stderr: 4: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x1a3) [0x7f7dafd0e85b]
2019-07-09T09:24:36.063 INFO:teuthology.orchestra.run.mira052.stderr: 5: (ceph::__ceph_assertf_fail(char const*, char const*, int, char const*, char const*, ...)+0) [0x7f7dafd0e9e5]
2019-07-09T09:24:36.064 INFO:teuthology.orchestra.run.mira052.stderr: 6: ceph-objectstore-tool() [0x1001c3d]
2019-07-09T09:24:36.064 INFO:teuthology.orchestra.run.mira052.stderr: 7: (BlueStore::_flush_cache()+0x80) [0x1061940]
2019-07-09T09:24:36.064 INFO:teuthology.orchestra.run.mira052.stderr: 8: (BlueStore::_fsck(bool, bool)+0x4ad9) [0x107b559]
2019-07-09T09:24:36.064 INFO:teuthology.orchestra.run.mira052.stderr: 9: (BlueStore::_mount(bool, bool)+0x20a) [0x1083aaa]
2019-07-09T09:24:36.064 INFO:teuthology.orchestra.run.mira052.stderr: 10: (main()+0x2dfe) [0xadc5fe]
2019-07-09T09:24:36.064 INFO:teuthology.orchestra.run.mira052.stderr: 11: (__libc_start_main()+0xf0) [0x7f7daeb15830]
2019-07-09T09:24:36.064 INFO:teuthology.orchestra.run.mira052.stderr: 12: (_start()+0x29) [0xaf21e9]
```

http://pulpito.ceph.com/kchai-2019-07-09_08:06:53-rados-wip-kefu-testing-2019-07-09-1127-distro-basic-mira/4105151/

markhpc commented:
New run results:

http://pulpito.ceph.com/nhm-2019-07-22_23:44:17-rados-wip-mark-testing-2019-07-22-0826-distro-basic-smithi/

tchaikov commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -880,454 +880,534 @@ int64_t BlueStore::GarbageCollector::estimate(
-  onode_lru.push_front(*o);
-}
+public:
+  TwoQBufferCacheShard(CephContext *cct) : BufferCacheShard(cct) {}
```
explicit.

tchaikov commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -880,454 +880,534 @@ int64_t BlueStore::GarbageCollector::estimate(
+      &BlueStore::Buffer::lru_item> > list_t;
+  list_t lru;
+
+  LruBufferCacheShard(CephContext *cct) : BlueStore::BufferCacheShard(cct) {}
```
explicit.

markhpc commented:
http://pulpito.ceph.com/nhm-2019-07-25_12:34:31-rados-wip-mark-testing-2019-07-24-0720-distro-basic-smithi/
http://pulpito.ceph.com/nhm-2019-07-25_00:46:08-rados-wip-mark-testing-2019-07-24-0720-distro-basic-smithi/
